src/geo.rs

```
@@ -0,0 +1,354 @@
+
+macro_rules! invalid_type_error {
+    ($v:expr, $det:expr) => {{
+        fail!((
+            ErrorKind::TypeError,
+            "Response was of incompatible type",
+            format!("{:?} (response was {:?})", $det, $v)
+        ));
+    }};
+}
+
+/// Units used by [`geo_dist`][1] and [`geo_radius`][2].
+///
+/// [1]: ../trait.Commands.html#method.geo_dist
+/// [2]: ../trait.Commands.html#method.geo_radius
+pub enum Unit {
+    Meters,
+    Kilometers,
+    Miles,
+    Feet,
+}
+
+impl ToRedisArgs for Unit {
+    fn write_redis_args<W>(&self, out: &mut W)
+    where
+        W: ?Sized + RedisWrite,
+    {
+        let unit = match *self {
+            Unit::Meters => "m",
+            Unit::Kilometers => "km",
+            Unit::Miles => "mi",
+            Unit::Feet => "ft",
+        };
+        out.write_arg(unit.as_bytes());
+    }
+}
+
+/// A coordinate (longitude, latitude). Can be used with [`geo_pos`][1]
+/// to parse response from Redis.
+///
+/// [1]: ../trait.Commands.html#method.geo_pos
+///
+/// `T` is the type of the every value.
+///
+/// * You may want to use either `f64` or `f32` if you want to perform mathematical operations.
+/// * To keep the raw value from Redis, use `String`.
+#[derive(Debug, PartialEq)]
+pub struct Coord<T> {
+    pub longitude: T,
+    pub latitude: T,
+}
+
+impl<T> Coord<T> {
+    /// Create a new Coord with the (longitude, latitude)
+    pub fn lon_lat(longitude: T, latitude: T) -> Coord<T> {
+        Coord {
+            longitude,
+            latitude,
+        }
+    }
+}
+
+impl<T: FromRedisValue> FromRedisValue for Coord<T> {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let values: Vec<T> = FromRedisValue::from_redis_value(v)?;
+        if values.len() != 2 {
+            invalid_type_error!(v, "Expect a pair of numbers");
+        }
+        let mut values = values.into_iter();
+        Ok(Coord {
+            longitude: values.next().unwrap(),
```

Unwrap here could panic. We can return an error instead.